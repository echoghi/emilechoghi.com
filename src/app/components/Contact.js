import React from 'react';
import { connect } from 'react-redux';
import { postForm, resetError } from './actions';
// Components
import { Helmet } from 'react-helmet';
import Footer from './Footer';
import Loading from './Loading';
import FormError from './Error';
import Button from '@material-ui/core/Button';
import Snackbar from '@material-ui/core/Snackbar';
import ContactMap from './ContactMap';
import ReactGA from 'react-ga';
import styled from 'styled-components';

const mapStateToProps = state => ({
    success: state.portfolioState.success,
    loading: state.portfolioState.loading,
    error: state.portfolioState.error,
    width: state.portfolioState.width
});

const mapDispatchToProps = dispatch => ({
    postForm: data => dispatch(postForm(data)),
    resetError: () => dispatch(resetError())
});

const Portfolio = styled.div`
    height: 80vh;
    padding: 0 70px;
    font-family: 'Varela Round';
    margin-top: 80px;

    @media (max-width: 1199px) and (min-width: 1024px) {
        height: auto;
        padding: 0 30px 30px 30px;
    }

    @media (max-width: 1023px) {
        height: auto;
        padding: 0;
        padding-bottom: 30px;
    }
`;

const FormHeader = styled.h4`
    font-size: 30px;
    font-weight: normal;
    margin: 100px 0;

    @media (max-width: 1199px) and (min-width: 1024px) {
        margin: 50px 0;
    }

    @media (max-width: 767px) {
        margin: 40px 0;
    }
`;

const Form = styled.form`
    font-family: 'Varela Round';
    background: none;
    margin: 0 auto;
    margin-top: 70px;
    width: 60%;

    @media (max-width: 1199px) and (min-width: 1024px) {
        height: 80vh;
        width: 70%;
    }

    @media (max-width: 1023px) {
        margin-top: 30px;
        width: 85%;
    }

    @media (max-width: 767px) {
        height: 185vh;
    }

    @media (max-width: 415px) and (min-width: 376px) {
        height: 100vh;
    }

    @media (max-width: 375px) {
        height: 105vh;
    }

    @media (max-width: 370px) {
        height: 120vh;
    }
`;

const FormRow = styled.div`
    display: block;
    width: 100%;
`;

const FormItem = styled.div`
    display: inline-block;
    margin: 10px 0;
    width: 46%;
    float: right;

    @media (max-width: 767px) {
        display: block;
        width: 100%;
    }

    &:first-child {
        float: left;
    }
`;

const FormItemLarge = styled.div`
    display: inline-block;
    margin: 10px 0;
    width: 100%;
    float: right;

    @media (max-width: 767px) {
        display: block;
        width: 100%;
    }

    &:first-child {
        float: left;
    }
`;

const ErrorLabel = styled.div`
    visibility: hidden;
    text-align: right;
    color: #dd4b39;
    letter-spacing: 0.3px;
    font-size: 12px;
    font-weight: bold;
    line-height: 2;
    text-transform: uppercase;

    &.invalid {
        visibility: visible;
    }
`;

const Label = styled.label`
    font-size: 18px;

    &.invalid {
        color: #dd4b39;
    }
`;

const Input = styled.input`
    font-family: 'Rubik', sans-serif;
    appearance: none;
    -moz-appearance: none;
    -webkit-appearance: none;
    display: block;
    box-sizing: border-box;
    background-color: #ecf1f6;
    color: #3d575d;
    border: none;
    width: 100%;
    margin: 5px 0;
    height: 55px;
    padding: 5px 0 5px 15px;
    font: inherit;
    font-size: 18px;
    border-radius: 0;
    transition: opacity 0.2s;

    &.invalid {
        border: 1px solid #dd4b39;
        border-radius: 0;
    }

    &:focus {
        outline: none;
    }
`;

const TextArea = styled.textarea`
    font-family: 'Rubik', sans-serif;
    appearance: none;
    -moz-appearance: none;
    -webkit-appearance: none;
    display: block;
    box-sizing: border-box;
    background-color: #ecf1f6;
    color: #3d575d;
    border: none;
    width: 100%;
    margin: 5px 0;
    height: 150px;
    padding: 15px 0 5px 15px;
    font: inherit;
    font-size: 18px;
    border-radius: 0;
    transition: opacity 0.2s;

    &.invalid {
        border: 1px solid #dd4b39;
        border-radius: 0;
    }

    &:focus {
        outline: none;
    }
`;

// Email validation RegExp
const validateEmail = /^(([^<>()\[\]\\.,;:\s@']+(\.[^<>()\[\]\\.,;:\s@']+)*)|('.+'))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
// Reusable validation constuctor for each input
let validationObj = function() {
    this.valid = false;
    this.dirty = false;
};

class Contact extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            loading: true,
            error: null,
            name: '',
            email: '',
            message: '',
            validation: {
                name: new validationObj(),
                email: new validationObj(),
                message: new validationObj()
            }
        };

        window.scrollTo(0, 0);
    }

    componentDidMount() {
        if (NODE_ENV === 'production') {
            ReactGA.ga('send', 'pageview', '/contact');
        }
    }

    renderLoading() {
        const { loading, error, success, resetError } = this.props;

        if (loading) {
            return <Loading />;
        } else if (error) {
            return <FormError close={resetError} />;
        }

        if (success && this.state.name) {
            this.resetForm();
        }
    }

    /**
     * Reset Form Data
     *
     * @state - entire state
     */
    resetForm() {
        // create a shallow copy of the state to mutate
        let obj = Object.assign({}, this.state);
        // Reset Form Data
        for (let key in obj) {
            if (key === 'validation') {
                obj[key].name = new validationObj();
                obj[key].email = new validationObj();
                obj[key].message = new validationObj();
            } else {
                obj[key] = '';
            }
        }

        // Reset inputs
        document.getElementById('contact-form').reset();

        this.setState(obj);
    }

    /**
     * Validate Inputs
     *
     * @return valid - validation status
     */
    validateInputs() {
        let valid = true;
        // Check for incompleted fields
        for (let key in this.state.validation) {
            if (!this.state.validation[key]['valid']) {
                return false;
            }
        }

        return valid;
    }

    /**
     * Validate/Mask inputs as the user types
     *
     * @param event - DOM event info
     * @state - Send validation status to state
     */
    onChange = event => {
        // create a shallow copy of the state to mutate
        let obj = Object.assign({}, this.state);
        // Set value in obj to eventually send to the state
        obj[event.target.name] = event.target.value;
        // Validate inputs
        // Mark input as dirty (interacted with)
        obj['validation'][event.target.name]['dirty'] = true;

        if (event.target.name === 'email') {
            // Validate email address
            if (validateEmail.test(event.target.value)) {
                obj['validation'][event.target.name]['valid'] = true;
            } else {
                obj['validation'][event.target.name]['valid'] = false;
            }
        } else {
            // If there is any value for non-email inputs, mark it valid
            if (event.target.value !== '') {
                obj['validation'][event.target.name]['valid'] = true;
            } else {
                obj['validation'][event.target.name]['valid'] = false;
            }
        }

        this.setState(obj);
    };

    /**
     * Handle Error classNames for each input
     *
     * @param name - Input name
     * @return className - return class depending on validation status
     */
    handleErrorClass = name => {
        let className;
        let validation = this.state.validation;

        if (validation[name].valid) {
            className = '';
        } else if (!validation[name].valid && validation[name].dirty) {
            className = 'invalid';
        } else if (!validation[name].valid && !validation[name].dirty) {
            className = '';
        }

        return className;
    };

    handleSubmit = event => {
        event.preventDefault();
        let { name, email, message } = this.state;

        if (this.validateInputs()) {
            this.props.postForm({ name, email, message });
        } else {
            // create a shallow copy of the state to mutate
            let obj = Object.assign({}, this.state);
            // If there is an invalid input, mark all as dirty on submit to alert the user
            for (let attr in this.state) {
                if (obj['validation'][attr]) {
                    obj['validation'][attr]['dirty'] = true;
                }
            }

            this.setState(obj);
        }
    };

    renderSnackbarStyle() {
        if (this.props.width < 760) {
            return {
                width: '100%'
            };
        }
    }

    render() {
        const { name, email, message } = this.state;

        return (
            <div>
                <Helmet>
                    <link rel="canonical" href="https://emilechoghi.com/contact" />
                    <link href="https://fonts.googleapis.com/css?family=Roboto" rel="stylesheet" />
                    <link
                        rel="stylesheet"
                        href="https://unpkg.com/leaflet@1.3.1/dist/leaflet.css"
                        integrity="sha512-Rksm5RenBEKSKFjgI3a41vrjkw4EVPlJ3+OiI65vTjIdo9brlAacEuKOiQ5OFh7cOI1bkDwLqdLw3Zg0cRJAAQ=="
                        crossOrigin=""
                    />
                </Helmet>
                <ContactMap />

                <Portfolio>
                    <div className="clearfix" />

                    <Form id="contact-form">
                        <FormHeader> Contact Me </FormHeader>
                        <FormRow>
                            <FormItem>
                                <Label className={this.handleErrorClass('name')}>Name</Label>
                                <Input
                                    type="text"
                                    name="name"
                                    value={name}
                                    maxLength="100"
                                    onChange={this.onChange}
                                    className={this.handleErrorClass('name')}
                                />
                                <ErrorLabel className={this.handleErrorClass('name')}>required*</ErrorLabel>
                            </FormItem>
                            <FormItem>
                                <Label className={this.handleErrorClass('email')}>Email Address</Label>
                                <Input
                                    type="text"
                                    name="email"
                                    value={email}
                                    maxLength="254"
                                    onChange={this.onChange}
                                    className={this.handleErrorClass('email')}
                                />
                                <ErrorLabel className={this.handleErrorClass('email')}>invalid*</ErrorLabel>
                            </FormItem>
                        </FormRow>
                        <FormRow>
                            <FormItemLarge>
                                <Label className={this.handleErrorClass('message')}>Message</Label>
                                <TextArea
                                    maxLength="6000"
                                    name="message"
                                    value={message}
                                    onChange={this.onChange}
                                    className={this.handleErrorClass('message')}
                                />
                                <ErrorLabel className={this.handleErrorClass('message')}>required*</ErrorLabel>
                            </FormItemLarge>
                        </FormRow>

                        <Button onClick={this.handleSubmit} className="form-button">
                            Send
                        </Button>
                    </Form>

                    {this.renderLoading()}

                    <Snackbar
                        open={this.props.success || false}
                        style={this.renderSnackbarStyle()}
                        message="Your message was sent, thanks for reaching out!"
                        autoHideDuration={4000}
                    />
                </Portfolio>

                <Footer fixed />
            </div>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Contact);
