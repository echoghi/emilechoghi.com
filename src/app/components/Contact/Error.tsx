import * as React from 'react';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogActions from '@material-ui/core/DialogActions';
import { Button } from './styles';

interface ErrorProps {
    close: () => void;
}

const Error: React.FunctionComponent<ErrorProps> = ({ close }) => (
    <Dialog open onClose={close}>
        <DialogTitle>Oops, your message wasn't sent ðŸ˜…</DialogTitle>
        <DialogContent>
            <DialogContentText>Something went wrong, please try again!</DialogContentText>
        </DialogContent>
        <DialogActions>
            <Button onClick={close}>Ok</Button>
        </DialogActions>
    </Dialog>
);

export default Error;
