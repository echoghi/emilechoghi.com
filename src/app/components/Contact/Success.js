import React from 'react';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContentText from '@material-ui/core/DialogContentText';
import { Button } from './styles';

const Success = ({ close }) => (
    <Dialog open onClose={close}>
        <DialogTitle>Thanks for reaching out! 👍</DialogTitle>
        <DialogContent>
            <DialogContentText>
                You should hear back from me within a day unless I'm on holiday or super busy.
            </DialogContentText>
        </DialogContent>
        <DialogActions>
            <Button onClick={close}>Ok</Button>
        </DialogActions>
    </Dialog>
);

export default Success;
