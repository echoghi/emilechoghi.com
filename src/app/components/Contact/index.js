import React, { useState, Fragment } from 'react';
import {
    TextArea,
    Input,
    Label,
    ErrorLabel,
    Form,
    FormHeader,
    FormItem,
    FormRow,
    FormItemLarge,
    Portfolio,
    FormButton
} from './styles';
// Components
import ReactGA from 'react-ga';
import Footer from '../Footer';
import Success from './Success';
import Loading from '../Loading';
import Error from './Error';

// Email validation RegExp
const validateEmail = /^(([^<>()\[\]\\.,;:\s@']+(\.[^<>()\[\]\\.,;:\s@']+)*)|('.+'))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

const Contact = () => {
    const [success, setSuccess] = useState(false);
    const [error, setError] = useState(false);
    const [loading, setLoading] = useState(false);
    // const [mounted, onMount] = useState(false);
    const [form, setFormValue] = useState({ name: '', email: '', message: '' });
    const [validation, setValidation] = useState({
        name: { dirty: false, valid: false },
        email: { dirty: false, valid: false },
        message: { dirty: false, valid: false }
    });

    // useEffect(
    //     () => {
    //         if (!mounted) {
    //             document.title = 'Contact Emile Choghi';

    //             if (NODE_ENV === 'production') {
    //                 ReactGA.ga('send', 'pageview', '/contact');
    //             }

    //             window.scrollTo(0, 0);
    //             onMount(true);
    //         }
    //     },
    //     [mounted]
    // );

    function renderLoading() {
        if (loading) {
            return <Loading />;
        }
    }

    function renderError() {
        if (error) {
            if (NODE_ENV === 'production') {
                ReactGA.event({
                    category: 'Form Error',
                    action: 'Form Submission Error',
                    label: 'Error Modal'
                });
            }

            return <Error close={() => setError(false)} />;
        }
    }

    function renderSuccess() {
        if (success) {
            return <Success close={() => setSuccess(false)} />;
        }
    }

    function resetForm() {
        // Reset Form Validation
        for (let type in validation) {
            validation[type] = { dirty: false, valid: false };
        }

        setFormValue({ name: '', email: '', message: '' });
        setValidation(validation);
    }

    function validateInputs() {
        // Check for incompleted fields
        for (let key in validation) {
            if (!validation[key]['valid']) {
                return false;
            }
        }

        return true;
    }

    const onChange = event => {
        const { name, value } = event.target;

        // Set value in obj to eventually send to the state
        form[name] = value;

        // Validate inputs
        if (name === 'email') {
            // Validate email address
            if (validateEmail.test(value)) {
                validation[name]['valid'] = true;
            } else {
                validation[name]['valid'] = false;
            }
        } else {
            // If there is any value for non-email inputs, mark it valid
            if (value !== '') {
                validation[name]['valid'] = true;
            } else {
                validation[name]['valid'] = false;
            }
        }

        setFormValue(form);
    };

    const handleErrorClass = name => {
        if (validation[name].valid) {
            return '';
        } else if (!validation[name].valid && validation[name].dirty) {
            return 'invalid';
        } else if (!validation[name].valid && !validation[name].dirty) {
            return '';
        }
    };

    const handleSubmit = event => {
        event.preventDefault();

        if (validateInputs()) {
            setLoading(true);

            return fetch('/functions/form', {
                method: 'POST',
                body: JSON.stringify(form),
                headers: { 'Content-Type': 'application/json; charset=utf-8' }
            })
                .then(response => {
                    if (response.status === 200) {
                        setSuccess(true);
                        setLoading(false);

                        if (NODE_ENV === 'production') {
                            ReactGA.event({
                                category: 'Form Success',
                                action: 'Message Submitted',
                                label: 'Success Notification'
                            });
                        }

                        resetForm();
                    } else {
                        setError(true);
                        setLoading(false);
                    }
                })
                .catch(error => {
                    setError(true);
                    setLoading(false);
                    throw error;
                });
        } else {
            // create a shallow copy of the state to mutate
            let obj = Object.assign({}, validation);
            // If there is an invalid input, mark all as dirty on submit to alert the user
            for (let attr in form) {
                if (obj[attr]) {
                    obj[attr]['dirty'] = true;
                }
            }

            setValidation(obj);
        }
    };

    const { name, email, message } = form;

    return (
        <Fragment>
            <Portfolio>
                <div className="clearfix" />

                <Form id="contact-form" onSubmit={handleSubmit}>
                    <FormHeader> Contact Me </FormHeader>
                    <FormRow>
                        <FormItem>
                            <Label className={handleErrorClass('name')}>
                                Name
                                <Input
                                    type="text"
                                    name="name"
                                    value={name}
                                    maxLength="100"
                                    onChange={onChange}
                                    className={handleErrorClass('name')}
                                    data-testid="name"
                                />
                            </Label>
                            <ErrorLabel
                                className={handleErrorClass('name')}
                                data-testid="name-label"
                            >
                                required*
                            </ErrorLabel>
                        </FormItem>
                        <FormItem>
                            <Label className={handleErrorClass('email')}>
                                Email Address
                                <Input
                                    type="text"
                                    name="email"
                                    value={email}
                                    maxLength="254"
                                    onChange={onChange}
                                    className={handleErrorClass('email')}
                                    data-testid="email"
                                />
                            </Label>
                            <ErrorLabel
                                className={handleErrorClass('email')}
                                data-testid="email-label"
                            >
                                invalid*
                            </ErrorLabel>
                        </FormItem>
                    </FormRow>
                    <FormRow>
                        <FormItemLarge>
                            <Label className={handleErrorClass('message')}>
                                Message
                                <TextArea
                                    maxLength="6000"
                                    name="message"
                                    value={message}
                                    onChange={onChange}
                                    className={handleErrorClass('message')}
                                    data-testid="message"
                                />
                            </Label>
                            <ErrorLabel
                                className={handleErrorClass('message')}
                                data-testid="message-label"
                            >
                                required*
                            </ErrorLabel>
                        </FormItemLarge>
                    </FormRow>

                    <FormButton type="submit" data-testid="send">
                        Send
                    </FormButton>
                </Form>

                {renderLoading()}
                {renderSuccess()}
                {renderError()}
            </Portfolio>

            <Footer fixed />
        </Fragment>
    );
};

export default Contact;
