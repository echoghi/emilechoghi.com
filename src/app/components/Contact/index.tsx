import React from 'react';
import {
    TextArea,
    Input,
    Label,
    ErrorLabel,
    Form,
    FormHeader,
    FormItem,
    FormRow,
    FormItemLarge,
    Portfolio,
    FormButton
} from './styles';
// Components
import ReactGA from 'react-ga';
import Footer from '../Footer';
import Success from './Success';
import Loading from '../Loading';
import Error from './Error';

interface ValidationState {
    email: { dirty: boolean; valid: boolean };
    message: { dirty: boolean; valid: boolean };
    name: { dirty: boolean; valid: boolean };
    [key: string]: { dirty: boolean; valid: boolean };
}

interface FormState {
    email: string;
    message: string;
    name: string;
    [key: string]: string;
}

// form validation
const formValidation: ValidationState = {
    email: { dirty: false, valid: false },
    message: { dirty: false, valid: false },
    name: { dirty: false, valid: false }
};

// form validation
const formState: FormState = {
    email: '',
    message: '',
    name: ''
};

// Email validation RegExp
const validateEmail: RegExp = /^(([^<>()\[\]\\.,;:\s@']+(\.[^<>()\[\]\\.,;:\s@']+)*)|('.+'))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

const Contact = React.memo(() => {
    const [success, setSuccess] = React.useState(false);
    const [error, setError] = React.useState(false);
    const [loading, setLoading] = React.useState(false);
    const [mounted, onMount] = React.useState(false);
    const [form, setFormValue] = React.useState(formState);
    const [validation, setValidation] = React.useState(formValidation);

    React.useEffect(
        () => {
            if (!mounted) {
                document.title = 'Contact Emile Choghi';

                if (NODE_ENV === 'production') {
                    // @ts-ignore
                    ReactGA.ga('send', 'pageview', '/contact');
                }

                window.scrollTo(0, 0);
                onMount(true);
            }
        },
        [mounted]
    );

    function renderLoading() {
        if (loading) {
            return <Loading />;
        }
    }

    function renderError() {
        if (error) {
            if (NODE_ENV === 'production') {
                ReactGA.event({
                    action: 'Form Submission Error',
                    category: 'Form Error',
                    label: 'Error Modal'
                });
            }

            const clearError = () => setError(false);

            return <Error close={clearError} />;
        }
    }

    function renderSuccess() {
        if (success) {
            const resetSuccess = () => setSuccess(false);

            return <Success close={resetSuccess} />;
        }
    }

    function resetForm(): void {
        setFormValue(formState);
        setValidation(formValidation);
    }

    function validateInputs(): boolean {
        // Check for incompleted fields
        for (const key in validation) {
            if (!validation[key].valid) {
                return false;
            }
        }

        return true;
    }

    const onChange = (event: any) => {
        const { value } = event.target;
        const formField = event.target.name;

        // Set value in obj to eventually send to the state
        form[formField] = value;

        // Validate inputs
        if (formField === 'email') {
            // Validate email address
            validation[formField].valid = validateEmail.test(value);
        } else {
            // If there is any value for non-email inputs, mark it valid
            if (value !== '') {
                validation[formField].valid = true;
            } else {
                validation[formField].valid = false;
            }
        }

        setFormValue(form);
    };

    const handleErrorClass = (inputName: string): string => {
        if (validation[inputName].valid) {
            return '';
        } else if (!validation[inputName].valid && validation[inputName].dirty) {
            return 'invalid';
        } else if (!validation[inputName].valid && !validation[inputName].dirty) {
            return '';
        }
    };

    const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
        event.preventDefault();

        if (validateInputs()) {
            setLoading(true);

            return fetch('/api/postForm', {
                body: JSON.stringify(form),
                headers: { 'Content-Type': 'application/json; charset=utf-8' },
                method: 'POST'
            })
                .then(response => {
                    if (response.status === 200) {
                        setSuccess(true);
                        setLoading(false);

                        if (NODE_ENV === 'production') {
                            ReactGA.event({
                                action: 'Message Submitted',
                                category: 'Form Success',
                                label: 'Success Notification'
                            });
                        }

                        resetForm();
                    } else {
                        setError(true);
                        setLoading(false);
                    }
                })
                .catch(err => {
                    setError(true);
                    setLoading(false);
                    throw err;
                });
        } else {
            // create a shallow copy of the state to mutate
            const obj = { ...validation };
            // If there is an invalid input, mark all as dirty on submit to alert the user
            for (const attr in form) {
                if (obj[attr]) {
                    obj[attr].dirty = true;
                }
            }

            setValidation(obj);
        }
    };

    const { name, email, message } = form;

    return (
        <React.Fragment>
            <Portfolio>
                <div className="clearfix" />

                <Form id="contact-form" onSubmit={handleSubmit}>
                    <FormHeader> Contact Me </FormHeader>
                    <FormRow>
                        <FormItem>
                            <Label className={handleErrorClass('name')}>
                                Name
                                <Input
                                    type="text"
                                    name="name"
                                    value={name}
                                    maxLength={100}
                                    onChange={onChange}
                                    className={handleErrorClass('name')}
                                    data-testid="name"
                                />
                            </Label>
                            <ErrorLabel
                                className={handleErrorClass('name')}
                                data-testid="name-label"
                            >
                                required*
                            </ErrorLabel>
                        </FormItem>
                        <FormItem>
                            <Label className={handleErrorClass('email')}>
                                Email Address
                                <Input
                                    type="text"
                                    name="email"
                                    value={email}
                                    maxLength={254}
                                    onChange={onChange}
                                    className={handleErrorClass('email')}
                                    data-testid="email"
                                />
                            </Label>
                            <ErrorLabel
                                className={handleErrorClass('email')}
                                data-testid="email-label"
                            >
                                invalid*
                            </ErrorLabel>
                        </FormItem>
                    </FormRow>
                    <FormRow>
                        <FormItemLarge>
                            <Label className={handleErrorClass('message')}>
                                Message
                                <TextArea
                                    maxLength={500}
                                    name="message"
                                    value={message}
                                    onChange={onChange}
                                    className={handleErrorClass('message')}
                                    data-testid="message"
                                />
                            </Label>
                            <ErrorLabel
                                className={handleErrorClass('message')}
                                data-testid="message-label"
                            >
                                required*
                            </ErrorLabel>
                        </FormItemLarge>
                    </FormRow>

                    <FormButton type="submit" data-testid="send">
                        Send
                    </FormButton>
                </Form>

                {renderLoading()}
                {renderSuccess()}
                {renderError()}
            </Portfolio>

            <Footer fixed={true} />
        </React.Fragment>
    );
});

export default Contact;
