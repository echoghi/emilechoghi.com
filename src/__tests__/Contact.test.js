import Contact from '../app/components/Contact';
import 'jest-styled-components';
import 'whatwg-fetch';

configure({ adapter: new Adapter() });

const mockFormPost = () => {
    return fetch('http://localhost:3000/api/postForm', {
        method: 'POST',
        body: JSON.stringify({ name: 'test', email: 'test@jest.com', message: 'testing API' }),
        headers: { 'Content-Type': 'application/json; charset=utf-8' }
    })
        .then(response => {
            if (response.status === 200) {
                return response.json();
            } else {
                console.log('POST test failed', response);
            }
        })
        .catch(error => {
            console.log('POST test failed', error);
        });
};

const component = mount(<Contact />);

describe('<Contact />', () => {
    test('snapshot', () => {
        expect(component).toMatchSnapshot();
    });

    test('form renders correctly', () => {
        // Test that component exists
        expect(component).toBeDefined();
        expect(component).toHaveLength(1);
    });

    test('form should submit', () => {
        // enter name
        const nameInput = component
            .find('input')
            .first()
            .hostNodes()
            .instance();
        nameInput.value = 'Test';

        // enter email
        const emailInput = component
            .find('input')
            .at(1)
            .hostNodes()
            .instance();
        emailInput.value = 'test@jest.com';

        // enter message
        const messageInput = component
            .find('textarea')
            .first()
            .hostNodes()
            .instance();
        messageInput.value = 'testing form';

        // submit form
        component.find('form').simulate('submit');

        // validation should pass
        expect(component.find('input').first()).toHaveStyleRule('border', 'none');
        expect(component.find('input').at(1)).toHaveStyleRule('border', 'none');
        expect(component.find('textarea').first()).toHaveStyleRule('border', 'none');
    });

    test('form validation should work', () => {
        const component = mount(<Contact />);

        // enter name
        const nameInput = component.find('input').first();
        nameInput.simulate('change', {
            target: { value: 'test', name: 'name' }
        });

        // enter email
        const emailInput = component.find('input').at(1);
        emailInput.simulate('change', {
            target: { value: 'broken email', name: 'email' }
        });

        // enter message
        const messageInput = component.find('textarea').first();
        messageInput.simulate('change', {
            target: { value: 'test', name: 'message' }
        });

        // submit form
        component.find('form').simulate('submit');

        // state should be updated
        expect(component.state('name')).toEqual('test');
        expect(component.state('email')).toEqual('broken email');
        expect(component.state('message')).toEqual('test');
        expect(component.state('validation')).toEqual({
            email: { dirty: true, valid: false },
            message: { dirty: true, valid: true },
            name: { dirty: true, valid: true }
        });
    });

    test('form should POST successfully', async function() {
        const response = await mockFormPost();

        expect(response.message).toBe('Email Sent');
    });
});
