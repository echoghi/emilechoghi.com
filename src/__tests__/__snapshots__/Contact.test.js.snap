// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Contact /> snapshot 1`] = `
<Fragment>
  <styled.div>
    <div
      className="clearfix"
    />
    <styled.form
      id="contact-form"
      onSubmit={[Function]}
    >
      <styled.h4>
         Contact Me 
      </styled.h4>
      <styled.div>
        <styled.div>
          <styled.label
            className=""
          >
            Name
          </styled.label>
          <styled.input
            className=""
            maxLength="100"
            name="name"
            onChange={[Function]}
            type="text"
            value=""
          />
          <styled.div
            className=""
          >
            required*
          </styled.div>
        </styled.div>
        <styled.div>
          <styled.label
            className=""
          >
            Email Address
          </styled.label>
          <styled.input
            className=""
            maxLength="254"
            name="email"
            onChange={[Function]}
            type="text"
            value=""
          />
          <styled.div
            className=""
          >
            invalid*
          </styled.div>
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.div>
          <styled.label
            className=""
          >
            Message
          </styled.label>
          <styled.textarea
            className=""
            maxLength="6000"
            name="message"
            onChange={[Function]}
            value=""
          />
          <styled.div
            className=""
          >
            required*
          </styled.div>
        </styled.div>
      </styled.div>
      <button
        className="form-button"
        type="submit"
      >
        Send
      </button>
    </styled.form>
  </styled.div>
  <Footer
    fixed={true}
  />
</Fragment>
`;
